### DISCUSSION TITLE
Transitioning from guest auth to user auth

### DISCUSSION THREAD

**Discussion by thardyman** [urql-graphql/urql#2855] [Q&A] [answered]
Hi, we're building out an app and initially we authenticate with our GraphQL server as a guest user.  If a user chooses to login, we have used [appauth-js](https://github.com/openid/AppAuth-JS) to handle the OpenID connect login flow and retrieve a new JWT and refresh token.  What is the best way to switch URQL to using the new tokens?

My thinking is that I set a flag "userHasLoggedIn" which I use in my `willAuthError` function.  If the user has just logged in, I'll make `willAuthError` return `true`.  This will force `getAuth` to run.  In `getAuth` I'll ignore the incoming `AuthState` and return the new `AuthState` that was retrieved from our OpenId Connect login flow.  I'll clear the "userHasLoggedIn" flag so that the future calls will continue to use the new, stored, `AuthState`.  

Does this make sense?  Is there a more elegant way to integrate with appauth-js?

**Comment by thardyman** [comment, marked as answer]
I've taken inspiration from this answer...
https://github.com/urql-graphql/urql/discussions/2647#discussioncomment-3501464

And used `mitt` to have a pub/sub model event listener.  It seems to work so far.  Although that answer seemed a bit incomplete.  Here's what I have.

Some global variables in my `authExchange.ts` module...

```Typescript
let hasBooted = false
let loggedInAuthState: AuthState | null = null

export const emitter = mitt()
```

I added the event listener, and the authState mutator into my `getAuth` function...

```Typescript
export const getAuth = async ({
  authState,
}: {
  authState?: AuthState | null
}): Promise<AuthState | null> => {
  if (loggedInAuthState) {
    console.log('Switching to logged in auth state')
    const returnVal: AuthState = {
      accessToken: loggedInAuthState.accessToken,
      refreshToken: loggedInAuthState.refreshToken,
      tokenExpiry: loggedInAuthState.tokenExpiry,
    }
    loggedInAuthState = null
    return Promise.resolve(returnVal)
  }
  if (!authState) {
    if (!hasBooted) {
      hasBooted = true
      emitter.on('login', (event: any) => {
        loggedInAuthState = {
          accessToken: event.accessToken,
          refreshToken: event.refreshToken,
          tokenExpiry: event.tokenExpiry,
        }
        console.log({ loggedInAuthState })
      })
    }
    const newAuthState = await renewAuthTokens()
    if (newAuthState) {
      return newAuthState
    }
  }

  clearRefreshToken()

  return null
}
```

And finally, I added the check for a `loggedInAuthState` into my `willAuthError` so that it would trigger my `getAuth` function...

```Typescript
export const willAuthError = ({
  authState,
}: {
  authState: AuthState | null
}): boolean =>
  Boolean(loggedInAuthState) ||
  (authState?.tokenExpiry ? authState.tokenExpiry - Date.now() < MIN_TTL : true)
```

It feels a bit messy having the global variables and event listener.  Is this how you intended it to be implemented @JoviDeCroock, or did I misunderstand your answer?

My app is a React app, so not sure if there is a React pattern that would better suit my needs 🤔 I looked at using the provider pattern, but I couldn't use hooks in the `getAuth` function so I got stuck with that approach.

**Reply by JoviDeCroock** [reply]
I mean, you don't really need the global variables, you can get the `loggedInAuthState` as part of `getAuth`, you can also register a logout event listener if you like. The event-emitter can mutate the `authState` removing the need for global variables.

To do it in the Provider you create the client as part of `useMemo` and the outer closure can define the authstate by means of a context/....
